src/main/drivers/serial_uart_stm32h7xx.c

```diff
@@ -0,0 +1,567 @@
+    }
+}
+
+#ifdef USE_DMA
+static void handleUsartTxDma(uartPort_t *s)
+{
+    uartTryStartTxDMA(s);
+}
+
+static void dmaIRQHandler(dmaChannelDescriptor_t* descriptor)
+{
+    uartPort_t *s = &(((uartDevice_t*)(descriptor->userParam))->port);
+    HAL_DMA_IRQHandler(&s->txDMAHandle);
+}
+#endif
+
+// XXX Should serialUART be consolidated?
+
+uartPort_t *serialUART(UARTDevice_e device, uint32_t baudRate, portMode_e mode, portOptions_e options)
+{
+    uartDevice_t *uartdev = uartDevmap[device];
+    if (!uartdev) {
+        return NULL;
+    }
+
+    uartPort_t *s = &(uartdev->port);
+
+    s->port.vTable = uartVTable;
+
+    s->port.baudRate = baudRate;
+
+    const uartHardware_t *hardware = uartdev->hardware;
+
+    s->USARTx = hardware->reg;
+
+#ifdef STM32H7
+    s->port.rxBuffer = hardware->rxBuffer;
+    s->port.txBuffer = hardware->txBuffer;
+    s->port.rxBufferSize = hardware->rxBufferSize;
+    s->port.txBufferSize = hardware->txBufferSize;
+#endif
+
+#ifdef USE_DMA
+    if (hardware->rxDMAStream) {
+        s->rxDMAStream = hardware->rxDMAStream;
+#if defined(STM32H7)
+        s->rxDMARequest = hardware->rxDMARequest;
+#else // F4 & F7
+        s->rxDMAChannel = hardware->DMAChannel;
+#endif
+    }
+
+    if (hardware->txDMAStream) {
+        s->txDMAStream = hardware->txDMAStream;
+#if defined(STM32H7)
+        s->txDMARequest = hardware->txDMARequest;
+#else // F4 & F7
+        s->txDMAChannel = hardware->DMAChannel;
+#endif
+
+        // DMA TX Interrupt
+        dmaIdentifier_e identifier = dmaGetIdentifier(hardware->txDMAStream);
+        dmaInit(identifier, OWNER_SERIAL_TX, RESOURCE_INDEX(device));
+        dmaSetHandler(identifier, dmaIRQHandler, hardware->txPriority, (uint32_t)uartdev);
+    }
+
+    s->txDMAPeripheralBaseAddr = (uint32_t)&s->USARTx->TDR;
+    s->rxDMAPeripheralBaseAddr = (uint32_t)&s->USARTx->RDR;
+#endif
+
+    s->Handle.Instance = hardware->reg;
+
+    IO_t txIO = IOGetByTag(uartdev->tx.pin);
+    IO_t rxIO = IOGetByTag(uartdev->rx.pin);
+
+    if ((options & SERIAL_BIDIR) && txIO) {
+        ioConfig_t ioCfg = IO_CONFIG(
+            ((options & SERIAL_INVERTED) || (options & SERIAL_BIDIR_PP)) ? GPIO_MODE_AF_PP : GPIO_MODE_AF_OD,
+            GPIO_SPEED_FREQ_HIGH,
+            ((options & SERIAL_INVERTED) || (options & SERIAL_BIDIR_PP)) ? GPIO_PULLDOWN : GPIO_PULLUP
+        );
+
+        IOInit(txIO, OWNER_SERIAL_TX, RESOURCE_INDEX(device));
+        IOConfigGPIOAF(txIO, ioCfg, uartdev->tx.af);
+    } else {
+        if ((mode & MODE_TX) && txIO) {
+            IOInit(txIO, OWNER_SERIAL_TX, RESOURCE_INDEX(device));
+            IOConfigGPIOAF(txIO, IOCFG_AF_PP, uartdev->tx.af);
+        }
+
+        if ((mode & MODE_RX) && rxIO) {
+            IOInit(rxIO, OWNER_SERIAL_RX, RESOURCE_INDEX(device));
+            IOConfigGPIOAF(rxIO, IOCFG_AF_PP, uartdev->rx.af);
+        }
+    }
+
+#ifdef USE_DMA
+#if defined(STM32H7)
+    if (!s->rxDMAStream)
+#else
+    if (!s->rxDMAChannel)
+#endif
+    {
+#endif
```

Comment:
If you move this to before the opening brace you can save the conditional around the closing brace.

Reply:
👍